### DISCUSSION TITLE
Does AD support operations except for the four rules?

### DISCUSSION THREAD

**Discussion by Raven-pro** [idaholab/moose#22648] [Q&A General] [answered]
Hello, everyone!

I know AD is based on the chain rule, and using AD variables and kernels can make automatically building jacobian a lot more convenient.

But I am wondering if AD support other operations like Exponential operations and Logarithmic operations and many other operations. 

For now I think exponential operations can be achieved using `Utility:pow<>` method (I dont know if I am right about that), but how about logarithmic operations and many others. Like, if just use `std::log` on an AD variable, can the derivative also be calculated correctly?

I am using MOOSE to develop an APP to solve single-phase flow problem, and the EOS requires operations mentioned above(IAPWS97 formulaitons). The properties are associated with main variables and I think the calculation process may affect the forming of jacobian(like setting them to be constants), increasing the errors, and leding to not convergening.

So is there any way to best implement proper operations like Exponential and Logarithmic?

Thank you very much!

**Comment by Raven-pro** [comment]
I just noticed that in the functions, if return type was set to ADReal, no error warning would show up when I use `MathUtils:pow` or `Utility::pow` to carry out Exponential operations.

Does that mean the derivative can also be properly computed by the both two ways?

**Comment by GiudGiud** [comment, marked as answer]
Hello

Yes AD is supported by most operators. I think atan2 or something similar might be the only one that got reported as not working. 

Will you be re-writing IAWPS properties? You need to have them written with ADReal inputs and return types (or DualReal, an AD number with less derivatives used in fluid properties)

Guillaume

**Reply by Raven-pro** [reply]
Thanks for your reply! 

Yes, I am considering adding some functions to the IAPWS properties right now.

I also have another question, is the derivative of an ADReal variable also an AD?

If that's so, when I want to compute some derivatives, I can just use `.derivative()[...]` to get that, am I right?

Thanks :)

**Reply by GiudGiud** [reply]
Which IAWPS properties? Do you have your own library?

You can get the derivatives through .derivative(), but they dont carry derivatives themselves. 
They rarely need to?

**Reply by cpgr** [reply]
Can I ask what is missing in IAPWS-IF97 (called `Water97FluidProperties`) that you need? It was a while since I wrote it, but I though that I had implemented everything in the IAPWS-IF97 paper.

**Reply by Raven-pro** [reply]
In fact,  I cannot find formulations of Region4 in `Water97FluidProperties`, like computing saturation values based on pressure and x, so I added related formulations in it. Except for that, almost all the variables are written in a non-AD manner.
I am trying to develop a single-phase water flow application, but the result is always wrong, so I am trying to find out the reason, like should I use AD on all the parameters like properties coefficients? I have no idea yet.
Is there any open-source single-phase MOOSE code available? I think maybe I need more references. :)

**Reply by Raven-pro** [reply]
> Can I ask what is missing in IAPWS-IF97 (called `Water97FluidProperties`) that you need? It was a while since I wrote it, but I though that I had implemented everything in the IAPWS-IF97 paper.

Hello, may I ask you about some convergence issues concerning single-phase 1D flow? I have opened another thread [#22682]. Could you please help me check if there is anything wrong with my code?  I just really cannot find out how to solve it alone, Thank you so much!

**Reply by cpgr** [reply]
Region 4 is the saturation curve - both saturation pressure as a function of temperature and saturation temperature as a function of pressure are implemented. There is a related IAPWS supplementary release for backward equations in Region 4 - is this what is missing?

**Reply by Raven-pro** [reply]
Yes, I didn't find related formulations in `Water97FluidProperties`, perhaps that's because I am using the old version MOOSE?
But the main problem I am facing is the convergence problem (that's why I tried to implement AD in fp), I am starting to think both AD and non-AD are okay concerning fluid properties as closures (Since my efforts have failed).
Could you please give me some suggestions about my convergence problem? THX!

**Reply by cpgr** [reply]
Yeah, it sounds like it could be something else

**Reply by GiudGiud** [reply]
@89018580 do you get convergence with constant fluid properties?

If there are some missing properties in `Water97FluidProperties` we would be very interested in that contribution

**Reply by Raven-pro** [reply]
> Yeah, it sounds like it could be something else

Oops, I just checked the `Water97FluidProperties.h` file, which I have not looked through carefully enough before.

The saturation temperature as the function of pressure is achieved in the `vaporTemperature` function. I thought there weren't formulations of Region4 because I didn't find "Region4" in `inRegion()` function, how careless I am.

The `Water97FluidProperties` does have everything in it.

**Reply by Raven-pro** [reply]
> @89018580 do you get convergence with constant fluid properties?
> 
> If there are some missing properties in `Water97FluidProperties` we would be very interested in that contribution

THX! Now I think `Water97FluidProperties` has everything, I was so careless.
I once thought the convergence problem was due to AD, as most variables in `Water97FluidProperties` are Real. Since in the kernels, terms like drho_dP and _drho_dT are also related to main variables P and H, so maybe they will contribute to Jacobian. So I tried to add AD in the functions, but it didn't make many differences. So I guess my convergence problem is not so related to AD issues.
If I set the properties like density to be constant, then according to the Mass Equation, the flux will also be constant, then the whole problem becomes a first-order problem considering p and h. It's an easy problem and can be solved, I have tried.
But the problem I am wanna solve is kind of a compressible flow problem, where fluid properties (water density mainly) are not constant.
I have read some papers, and I think MOOSE can definitely solve the 1D compressible problem, but when I ran the application, it just could not get the right result, yet could not converge ultimately. I wonder if it's an issue with my physical models or if I have done something wrong owing to my limited knowledge about MOOSE, I just don't know how to fix that.
Could you please have a look at my code and give me some suggestions?
The specific description of my problem is in thread [#22682].

**Reply by GiudGiud** [reply]
So the AD properties are provided, basically there is a macro that creates the AD property from the regular (Real) property + the derivatives routine. 

In your calls to these properties, you need to make sure you call the AD version though. The inputs to the fluid properties calls should be AD variables, and the output should be caught in an ADReal / DualReal as well. 

I would really recommend you look at the thermal hydraulics module. We support compressible flow there iirc (@joshuahansel )

**Reply by joshuahansel** [reply]
Yes, the MOOSE thermal hydraulics module (https://mooseframework.inl.gov/modules/thermal_hydraulics/index.html) has a 1D compressible flow model for you already. It is equivalent to an HLLC finite volume scheme with or without slope reconstruction.

**Reply by Raven-pro** [reply]
> Yes, the MOOSE thermal hydraulics module (https://mooseframework.inl.gov/modules/thermal_hydraulics/index.html) has a 1D compressible flow model for you already. It is equivalent to an HLLC finite volume scheme with or without slope reconstruction.

Yeah, I have looked into the thermal hydraulics module. I see all the kernels needed are encapsulated in the `Components`. 
I have read the `FlowChannel1Phase.C` code and `FlowSinglePhase.C` code, and in line 271 the code was adding moose kernels into the App, but there weren't advection terms. 
Were advection terms omitted from the TH module?

**Reply by joshuahansel** [reply]
> Were advection terms omitted from the TH module?

The advection terms are added via DG kernels. Search for "ADNumericalFlux3EqnDGKernel" in `FlowModelSinglePhase.C`.

**Reply by Raven-pro** [reply]
> > Were advection terms omitted from the TH module?
> 
> The advection terms are added via DG kernels. Search for "ADNumericalFlux3EqnDGKernel" in `FlowModelSinglePhase.C`.

Oh！ Thanks a lot! DGKernel is totally a new thing to me, is there any tutorial pages about it? THX!

**Reply by joshuahansel** [reply]
I don't know about tutorial pages, but this is the doc page: https://mooseframework.inl.gov/syntax/DGKernels/index.html

**Reply by Raven-pro** [reply]
> I don't know about tutorial pages, but this is the doc page: https://mooseframework.inl.gov/syntax/DGKernels/index.html
Thank you, I also want to ask one minor question, why do you use DGKernels instead of ordinary Kernels system as the Advection terms?
I think this may is the point that I am missing and why my calculation cannot get right result. :)

**Reply by joshuahansel** [reply]
Kernels operate on elements, while DGKernels work on internal sides. The choice of which to use depends on your spatial discretization.

Take this scalar advection equation as an example:
```
du/dt + df/dx = 0
```
If using CG, you'd multiply by your test function, integrate over the domain, and integrate the advection terms by parts:
```
(du/dt, phi_i) - (f, grad_phi_i) + <f, phi_i> = 0
```
The first two terms are kernels, and the last term will be a BC.

However, if doing finite volume, as done in THM, you just integrate over a cell:
```
(du/dt, 1) + f_{i+1/2} - f_{i-1/2} = 0
```
so the first term is a kernel again, but the flux terms are DG kernels.

**Reply by Raven-pro** [reply]
Thank you! That will be of great help!
